library/std/src/io/mod.rs

```diff
@@ -430,6 +430,7 @@ pub(crate) fn default_read_to_end<R: Read + ?Sized>(
         loop {
             match r.read(&mut probe) {
                 Ok(n) => {
+                    buf.try_reserve(n).map_err(|_| ErrorKind::OutOfMemory)?;
                     buf.extend_from_slice(&probe[..n]);
```

Comment:
This looks like a problem. It reads first and then it tries to allocate. When that fails we drop the read bytes on the floor. Which means the error is not recoverable. So I think we have to keep panicking here. Which is probably ok because the process is in dire straits anyway when it can't satisfy a tiny allocation request.

But that means we can't promise this for the default implementation.

Reply:
Indeed. I've reverted the documentation promise, removed this `try_reserve`, and updated the example code to warn about this gotcha.